[WIP] Feature/blueprint clone command

walreyes opened the pull request:
### What does this PR do?
* Adds `blueprint clone <blueprint-name> <name> [options]` command.
* Solves #108.

### TODO
 
* [x] Review from maintainers.
* [ ] Replace cloneTo path to the new `settings.cloneTo`
* [ ] Review Blacklisted Files
* [ ] Add Specs.

walreyes committed:
Adds BlueprintCloner model

Why is this change necessary?
  For the clone command there needs to be a model to clone files.

How does this address the issue?
  BlueprintCloner is initialized with a blueprint and some options.
  Then it implements a `clone()` method that will clone all the blueprint files.
  It ignores blacklisted files (.ds_store, .git, .gitkeep))

walreyes committed:
Adds Sub-command Clone class.

Sub-command Clone simplements a `run(blueprintName, args)` that simply calls the task `CloneBlueprint` to clone a blueprint.

walreyes committed:
Adds CloneBlueprint task.

CloneBlueprint is initialized with the blueprint that needs to be cloned and some options for the destination blueprint.
CloneBlueprint calls BlueprintCloner to clone the files

walreyes committed:
Adds clone cli command

Implements a slightly different version of build-blueprint-commands and build-blueprint-command.

If the cli will have more commands it might be a good idea to make those two files adaptable to different descriptions, usages, etc.

walreyes commented:
Hey @anithri!

I have'nt added the specs because I first want to know if this looks good to you:

#### The Process I followed to add a new command:

I followed the same structure `generate` has:

1.  I added the `clone cli command` under `cli/cmds/clone.js` and implemented a slightly different version of `buildBlueprintCommands` and `buildBlueprintCommand` (if the cli will have more commands these 2 classes could be refactored to be able to build different subcommands).
2.  Added the Sub-Command `Clone` class and `CloneBlueprint` task.
3.  Added a new model `BlueprintCloner` to clone the blueprint files into a new destination.

Each commit has a more comprehensive explanation of what each added class does.

#### Questions

Is there a correct way to get the `cloneTo` path right now? 
I am doing:  `settings.blueprints.searchPaths[0]`  but I'm not sure if it's correct.

anithri commented on src/models/blueprint-cloner.js:
```diff
@@ -0,0 +1,84 @@
+const FILES_BLACKLIST = ['.ds_store', '.git', '.gitkeep'];
+
+export default class BlueprintCloner {
+  constructor(blueprint, options) {
```
let's use `constructor(blueprint, options = {})` instead of line 12

walreyes replied:
Done!

anithri commented on src/cli/cmds/clone.js:
```diff
@@ -0,0 +1,35 @@
+
+module.exports = {
+  command: 'clone <blueprint> <name>',
+  aliases: ['g', 'gen'],
```
probably the wrong aliases

jamesr73 replied:
indeed... aliases are optional so you can set an empty array or remove the key entirely.

walreyes replied:
Thanks! 

I've fixed it.

anithri commented on src/cli/cmds/clone.js:
```diff
@@ -0,0 +1,35 @@
+      .usage(usage)
+      .option('dry-run', {
+        alias: 'd',
+        describe: "List files but don't generate them",
```
~2 different quoting styles used in this line and last/next.  If you install and run prettier, that will help with your formatting.~

~This is the only descepency that eslint finds is in switch statements. Prettier wants cases indented, and eslint wants them flush with the switch statement.~
```
// NO
export default function(state = defaultState, action) {
  switch (action.type) {
    default:
      return state;
  }
}
// YES
export default function(state = defaultState, action) {
  switch (action.type) {
  default:
    return state;
  }
}
```
~And so I've been editing it back after i `yarn run pretty`~

jamesr73 replied:
@anithri 

Prettier would use the two different quoting styles, because the describe string has an embedded single quote. Eslint should already allow that (and complain if double quotes are used in any other circumstance). The rule for quoting was modified by a recent PR

`"quotes": [2, "single", {"avoidEscape": true}],`

I prefer the prettier formatting for switch, and not having any manual step to undo it! This can also be accommodated by eslint:

`"indent": [2, 2, {"SwitchCase": 1}],`

Suggest we make that change in `.eslintrc`

anithri commented on src/models/blueprint-cloner.js:
```diff
@@ -0,0 +1,84 @@
+
+import { fileExists } from '../util/fs';
+
+const FILES_BLACKLIST = ['.ds_store', '.git', '.gitkeep'];
```
We definitely need a `FILES_BLACKLIST`.  I'm just wondering if there's value to making it an rc option that would include what you have here as defaults.

walreyes replied:
It might be worth it 🤔 

If we set the defaults it's always good to let the user add more customization.

But in what other circumstances will we need to blacklist files besides `clone`?

anithri commented on src/models/blueprint-cloner.js:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  clone() {
+    const blueprint  = this.blueprint;
```
By the time a blueprint is added to the collection, it will already have a validated existent path to itself.  And will also have a method that returns it's files.

walreyes replied:
Awesome!

I'll make the changes once the new api is ready.

anithri commented on src/models/blueprint-cloner.js:
```diff
@@ -0,0 +1,84 @@
+    const blueprint  = this.blueprint;
+    this.ui.writeInfo('cloning blueprint...');
+
+    const cloneToDirectory = this.cloneToDirectory();
```
The directory we clone to will be determined at the settings level, something like `rc.get('path.bpTarget`). So by this point in execution, you'll either have a falsy value or a validated, existing path you can write to.  Falsy values should generate a warning or an error.  

It seems to me that making assumptions based on the search path order, is risking unintended consequences.  Better to get 'permission' with an explicity defined path.

walreyes replied:
I totally agree, it is risky to make assumptions based on the search path order and it does make sense to get that `permission` from the explicit path.

I was aware that I will need to change the `cloneDirectory` to something like `rc.get('path.bpTarget')`. 

I will make the change once the new `rc` api is ready.

anithri commented on src/models/blueprint-cloner.js:
```diff
@@ -0,0 +1,84 @@
+    return path.resolve(settings.blueprints.searchPaths[0], this.newBlueprintName());
+  }
+
+  newBlueprintName() {
```
I think this is another thing that should be from `rc`, so that we can add `--name <newName>` or `--cloneTo` or similar.

walreyes replied:
This is a great idea.

Let me check if I am understanding correct, rc will have:

* All the settings from the `.blueprintrc`.
* The UI environment to write.
* The arguments sent from the CLI.

anithri commented on src/tasks/clone-blueprint.js:
```diff
@@ -0,0 +1,32 @@
+import Task from '../models/task';
+import BlueprintCloner from '../models/blueprint-cloner';
+
+export default class extends Task {
```
Lets add an explicit class name, if only to make our IDE's happier.

jamesr73 replied:
FYI... none of the other Tasks currently specify a class name.

anithri replied:
well, that's true.  AirBNB recommend the existing way too: 
https://github.com/airbnb/javascript#modules--prefer-default-export

TIL

anithri commented on src/cli/cmds/clone/build-blueprint-commands.js:
```diff
@@ -0,0 +1,26 @@
+    _cloneDeep(bp[blueprint.name])
+  ));
+
+const buildBlueprintCommands = () => {
```
All of the blueprints will already be found, read,  and have it's settings already merged.  Again the api is not quite ready yet, but soon.  And any suggestions would be most welcome. see #132 for current thinking

anithri commented on src/cli/cmds/clone.js:
```diff
@@ -0,0 +1,35 @@
+  command: 'clone <blueprint> <name>',
+  aliases: ['g', 'gen'],
+  describe: 'Clone a blueprint with a different name',
+  builder: yargs => {
```
We'll want all commands to follow the same pattern, I'm working to finalize the api for that, but take a look at #132 for the current thinking. @jamesr73 ?

jamesr73 replied:
I'll take a closer look at this before providing final feedback... this seems to fall between the pattern used for simple yargs commands like init and the more involved pattern used to support the blueprint specific generate commands.

anithri reviewed (commented):
This looks good so far.  Most of the really big issues are due to upstream api's not being finalized yet.

jamesr73 reviewed (commented):
The clone logic looks good. The CLI needs handlers similar to:

```
// src/cli/cmds/generate

...
getHandler().onRun('generate', handlers.handleRun);
getHandler().onHelp('generate', handlers.handleHelp);
...
```

These are used by generate so it can parse argv in 2 stages, the first pass gets enough information to setup the environment and load the available blueprints, the second pass then runs the specific blueprint generate command, or outputs help. 

Note that for the handlers to work `generate.js` `module.exports` does not include a `handler` property/function. And the `builder` property/function configures `yargs` with `.exitProcess(false)` and `.fail(fn)`.

It's not clear that the clone command needs access to the blueprint specific options (yet?) so it could be a single command handler rather than needing to buildBlueprintCommands. I'm open to discussion on that.

I think the basic CLI interface would be:

`bp clone <blueprint> [new_name]`

Which would copy <blueprint> from wherever it is found on the search path to the local directory, optionally setting a new name. This would also support cloning a blueprint that is already in the local directory to a new name.

The other thing to consider is the help output:

`bp help clone`

walreyes committed:
Solves clone wrong aliases

walreyes committed:
Runs pretty in blueprint-cloner.js

walreyes committed:
Defaults options = {} in BlueprintCloner constructor

walreyes commented:
Hey @anithri: 

I see 3 important changes:

* [ ] Change the `cloneToDirectory` to the one provided by the new `rc`
* [ ] Add the `--new-name` and `--clone-to` options from the CLI
* [ ] Remove blueprint files location logic and get those files from `blueprint.files()` 

1 change that we need to discuss if it's worth changing:

* [ ] Move blacklist files to rc and set defaults to: `['.ds_store', '.git', '.gitkeep']`.

walreyes commented:
Hey @jamesr73 

I think you are right, `clone` it's not as complex as `generate` so it might not need the blueprint options. I don't think it needs them unless we think `clone` could do more besides cloning the blueprint files.

So I will make the change to not build all blueprint commands but only build the clone command 
 which will run the `clone-blueprint` task which will then look for the `blueprint` in the `blueprint-collection` and the send it over to the `BlueprintCloner`.

I am not sure yet how the generate `handlers` work 🤔  but it's only a matter of digging into the code to figure it out.

I'll refactor the clone command with the handlers.

walreyes commented:
Thanks for the feedback @jamesr73 @anithri!

I will be making the changes I can without the new `rc` and then I'll add the remaining once it's ready.

jamesr73 commented:
Hey @walreyes 

The handlers are just callbacks registered with EventEmitters that are triggered after `yargs` has parsed the command line. See `src/cli/index.js` and `src/cli/handler.js`. This mechanism was added to support the `rc` api changes that are in progress. In particular it allows the `yargs` built in `help`  support to be extended with information read from the settings.

walreyes committed:
Refactors clone cmd to a simpler setup.

Why is this change necessary?
  I used buildBlueprintCommands but `clone` does'nt really needs the blueprint options, with the name it's enough to clone the files

How does this address the issue?
  Deletes clone/buildBlueprintCommands and uses handles to handle run

walreyes commented:
@jamesr73 got it. So `yargs` emits an event with the command name and then the handlers pick them up and run the proper command. It's awesome!

I've refactored the `cli/cmds/clone.js` to a much simpler setup, a middle ground between `init, config, new` and `generate`.

What do you think?

jamesr73 commented:
@walreyes looks good, much cleaner and clearer. I think we should also add a handler for help. Yargs will already provide it's default help for the command name, usage and options. The help handler should list the available blueprints. It doesn't need to use yargs to generate this, just log to the console (or use the ui).

jamesr73 commented on src/cli/cmds/clone/handlers.js:
```diff
@@ -0,0 +1,25 @@
+
+export default handlers;
+
+function handleRun(argv, yargs, rawArgs = process.argv.slice(3)) {
```
The `rawArgs` parameter can be removed as it isn't needed here in the same way as it was for `generate`.

jamesr73 commented on src/cli/cmds/clone/handlers.js:
```diff
@@ -0,0 +1,25 @@
+  const blueprintName = argv.blueprint;
+  const environment = getEnvironment();
+
+  if (blueprintExists(environment.settings.blueprints, blueprintName)) {
```
Pedant suggestion...

```javascript
if (blueprintInEnvironment(blueprintName, environment) {
...
}
```
